src/Terrain/Environment.java

```diff
@@ -1,49 +1,29 @@
 package Terrain;
 
-import java.util.ArrayList;
-
-import Terrain.Types.Airport;
-import Terrain.Types.BaseTerrain;
-import Terrain.Types.Bridge;
-import Terrain.Types.City;
-import Terrain.Types.Dunes;
-import Terrain.Types.Factory;
-import Terrain.Types.Forest;
-import Terrain.Types.Grass;
-import Terrain.Types.Headquarters;
-import Terrain.Types.Lab;
-import Terrain.Types.Mountain;
-import Terrain.Types.Reef;
-import Terrain.Types.Road;
-import Terrain.Types.Sea;
-import Terrain.Types.Seaport;
-import Terrain.Types.Shoal;
+import java.util.HashMap;
+import java.util.Map;
 
 /**
  * Environment is a flyweight class - each Terrain/Weather combination is instantiated only once.
  * Subsequent calls to retrieve that tile will receive the same copy.
  */
 public class Environment
 {
-  
-  private static ArrayList<BaseTerrain> terrainTypes;
-
   public enum Weathers
   {
     CLEAR, RAIN, SNOW, SANDSTORM
   };
 
-  public final BaseTerrain terrainType;
+  public final TerrainType terrainType;
   public final Weathers weatherType;
 
-  // ArrayList can dynamically expand to accomodate arbitrary terrains at runtime
-  // Thankfully, there's no plans to expand our list of weathers
-  private static ArrayList<Environment[]> tileInstances= new ArrayList<Environment[]>();
+  // Maintain a list of all tile types. Each type will be added the first time it is used.
+  private static Map<TerrainType, Environment[]> tileInstances= new HashMap<TerrainType, Environment[]>();
```

Comment:
Yeah, a Map is better than an Arraylist for this. Guess I'm just not used to having them around yet.